### PR TITLE
Major UIKit Rendering Performance Improvements, Logo Customization, & LitElement Migration

### PR DESCRIPTION
**Pull request by sghoweri**
- Significant rendering performance improvements when resizing the Drawer + Iframe resizer
- Added the ability to auto-close the Nav + Context menu via hitting the `esc` key
- Added the ability to auto-close the Nav + the Context menu via clicking anywhere outside of either menu — even inside the iframe!

- Improved (and slowly, more consistent) focus styles for navigating the Pattern Lab UI via Keyboard

CleanShot 2019-10-20 at 14 24 43

- Updated the SVG icons used in UIKit to automatically inherit the parent's text color via `currentColor`; added the light / dark theme + horizontal / vertical layout SVG icons that were previously inlined into the component-specific JS

- Converted several of our web components' internal JS rendering to use lit-html vs previously using Preact

- Added a brand new `<pl-icon>` web component (powered by lit-element!) + updated our Webpack build process to auto-generate an SVG spritesheet based off of the icons available in the UIKIt `src/icons` folder.
  - Updated several components to take advantage of this new Icon component

- Added the ability to customize the UIKit logo! This includes the ability to customize:
  - The image you use on both light and dark themes
  - Logo url
  - Logo text (including hiding this text and just displaying a logo)
  - Logo alt text
 
80047293-26FA-40E6-9049-0AF683463726

Closes #
- #1055 (partially)

### CODE REVIEW

**Comment by bradfrost** [comment]
@sghoweri thanks for this work! Hey, so if possible I'd like to tighten a few things up with the UIKit from a style point of view. What would be the best way to coordinate with you on that?

**Comment by sghoweri** [comment]
> @sghoweri thanks for this work! Hey, so if possible I'd like to tighten a few things up with the UIKit from a style point of view. What would be the best way to coordinate with you on that?

Hey @bradfrost!

Is this in regards to this particular PR or just Pattern Lab’s UI in general? I’m assuming you mean in general but just confirming!

As far as coordinating UI changes go, whatever is easiest for you to communicate these UI changes works for me (literally making CSS changes yourself, putting together a list. annotating screenshots, spinning up a PL instance for Pattern Lab’s UI itself), etc. Just a head’s up that there are still quite a number of things** on my radar so a lot of code is still in flux 😉.

In any case, I’d love to tag team on this. LMK!

  
**At a high level these mostly include:  
- More customization options (ex. customize the logo on light / dark themes, URL, etc; deep color palette customizations via CSS custom properties)
- More consistent behavior (ex. more consistent Nav behavior between smaller and larger screens — less hacks and overrides!)
- Accessibility (larger tappable target areas, better keyboard support, deep support for OS preferences to reduce motion / higher contrast, decent UI contrast, more work improving our “no-JS” progressive enhancement, etc) 
- UI Consistency (consolidate duplicate UI like Buttons, Icons, Dropdowns, Lists, etc)
- Internal refactoring (internally switch over web component rendering from using Preact and LitHTML to LitElement; upgrade approach to loading component-level CSS, conditionally switch on Shadow DOM encapsulation)
- Overall code cleanup and refactoring — especially work to more easily support end user customizations to the theming system / UI density via CSS custom properties
- And finally docs / automated testing coverage. 🙂

**Comment by bmuenzenmeyer** [comment]
fixed the preview deploy so it no longer is failing - but i do see this now:

image

**Comment by sghoweri** [comment]
> fixed the preview deploy so it no longer is failing - but i do see this now:
> 
> image

Thanks @bmuenzenmeyer! Saw that — probably just a default value I need to fix.

Also FWIW,  I’m actually thinking about getting a handful of Jest tests in (either with this PR or as a fast follow) to get the ball rolling on component / feature level testing... 🤔 #fewerSurprises

**Comment by sghoweri** [comment]
@bmuenzenmeyer lol found it...😅

https://github.com/pattern-lab/patternlab-node/commit/c95a06ece78631b068f8721666caf33452e57a7a

**Comment by bradfrost** [comment]
> Is this in regards to this particular PR or just Pattern Lab’s UI in general? I’m assuming you mean in general but just confirming!

Yep, in general! Sorry for hijacking the PR thread to ask, but since it's UI related I decided to ask.

> As far as coordinating UI changes go, whatever is easiest for you to communicate these UI changes works for me (literally making CSS changes yourself, putting together a list. annotating screenshots, spinning up a PL instance for Pattern Lab’s UI itself), etc. Just a head’s up that there are still quite a number of things** on my radar so a lot of code is still in flux 😉.

Awesome, that's super helpful thanks! I think what I'll do is make a branch, tweak some things, and then we can get together to walk through them. I don't have anything super major in mind, but just some tightening up/tweaky things. Sound like a plan?

**Comment by sghoweri** [comment]
> > Is this in regards to this particular PR or just Pattern Lab’s UI in general? I’m assuming you mean in general but just confirming!
> 
> Yep, in general! Sorry for hijacking the PR thread to ask, but since it's UI related I decided to ask.
> 
> > As far as coordinating UI changes go, whatever is easiest for you to communicate these UI changes works for me (literally making CSS changes yourself, putting together a list. annotating screenshots, spinning up a PL instance for Pattern Lab’s UI itself), etc. Just a head’s up that there are still quite a number of things** on my radar so a lot of code is still in flux 😉.
> 
> Awesome, that's super helpful thanks! I think what I'll do is make a branch, tweak some things, and then we can get together to walk through them. I don't have anything super major in mind, but just some tightening up/tweaky things. Sound like a plan?

👍🏼 Sounds great!

**Comment by sghoweri** [comment]
I need to update the PR description on this + add a few more screenshots but other than that I think this should be ready to go!

Super excited with how this turned out!!

**Comment by sghoweri** [comment]
@bmuenzenmeyer I couldn't help but notice that the most recent Netlify deploy on this branch seems to be missing the Handlebars data (but the UIKit still seems to be rendering just fine without any errors)... any ideas?

Pulling up a fresh copy / build of this locally seems to be working fine (testing in the `edition-twig` package) so I'm wondering if that's just a Netlify fluke.. 🤔

**Comment by bmuenzenmeyer** [comment]
@sghoweri if you could, please feel free to merge and release - I am about to step out the door and will be out for a while, so I dont think it smart to release and then bail.

File: packages/uikit-workshop/src/scripts/utils/polyfills/polyfill-loader.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ export const loadPolyfills = new Promise(resolve => {
     // Based on https://github.com/webcomponents/webcomponentsjs/blob/master/entrypoints/webcomponents-hi-ce-index.js
     // Used in: Safari 10, Firefox once SD is shipped
     Promise.all([
-      import('@webcomponents/shadydom/src/shadydom.js'),
+      // import('@webcomponents/shadydom/src/shadydom.js'),
```
[the remark this answers is not in the document]
Comment (sghoweri) [reply packages/uikit-workshop/src/scripts/utils/polyfills/polyfill-loader.js]: Yep - that’s on my todo list. 

Already cleaned these up in my local version of these updates 🧽

**Comment by sghoweri** [comment]
> @sghoweri if you could, please feel free to merge and release - I am about to step out the door and will be out for a while, so I dont think it smart to release and then bail.

Thanks @bmuenzenmeyer!!

I’m good waiting till tomorrow (or even Thursday) before I merge all this down and cut a new release. Want to make sure I keep an eye on the NPM publish + see if I can squeeze in some markdown docs to help focus configure the new Logo options.

**Comment by bmuenzenmeyer** [comment]
> see if I can squeeze in some markdown docs to help focus configure the new Logo options.

or PR it into patternlab.io? I know Brad is working on a new design but adding more content shouldn't get too much in the way like I did for https://github.com/pattern-lab/website/issues/121

**Pull request merged**